DashboardView data display implemented

quiz3 opened the pull request:
DashboardView data display implemented with two new tables through which the user is given information. The tables are all shown, the titles are displayed where appropriate, the information is accessible to the user, and all the buttons work. 

This will likely lead to merge conflicts because many changes have been made to main since the branch was created, so please bear with me and I can help if need be with the resolutions to those issues. 

EDIT: I have resolved the merge conflicts.

quiz3 committed:
DashboardView data display implemented with two new tables with which the user is given information

awesominat commented on src/main/java/interface_adapters/Dashboard/DashboardState.java:
```diff
@@ -1,14 +1,16 @@
 package interface_adapters.Dashboard;
 
-import entities.PortfolioInformation;
+import java.util.ArrayList;
```
nice

quiz3 replied:
What's nice about it? I don't even remember what the point of that change was. Re-ordering imports I imagine.

awesominat replied:
decoupling from the entity class is a really good addition

awesominat commented on src/main/java/view/DashboardView.java:
```diff
@@ -81,93 +96,201 @@ public DashboardView(DashboardViewModel dashboardViewModel,
-                        }
-                    }
+                evt -> {
+                    DashboardState currentState = dashboardViewModel.getState();
```
`currentState` is not used

awesominat commented on src/main/java/view/DashboardView.java:
```diff
@@ -81,93 +96,201 @@ public DashboardView(DashboardViewModel dashboardViewModel,
 
-                        }
-                    }
+                    // TODO: add reset balance pop-up
```
I dont think we are doing the pop-up. add it as an issue

quiz3 replied:
So you're saying I should remove the comment and then create an issue for it... ?

awesominat commented on src/main/java/view/DashboardView.java:
```diff
@@ -81,93 +96,201 @@ public DashboardView(DashboardViewModel dashboardViewModel,
+        // Spice up the actual balance (field) as well.
+        Font fieldFont = balanceField.getFont();
+        balanceField.setForeground(Color.DARK_GRAY);
+        balanceField.setFont(fieldFont.deriveFont(Font.ITALIC));
```
nice font usage

quiz3 replied:
Thanks! I spent a little too long figuring that out :)

awesominat commented on src/main/java/view/DashboardView.java:
```diff
@@ -81,93 +96,201 @@ public DashboardView(DashboardViewModel dashboardViewModel,
+        DefaultTableModel tableModel = new DefaultTableModel();
+
+        // Add columns to the model
+        tableModel.addColumn("Tickers");
+        tableModel.addColumn("Full Names");
+        tableModel.addColumn("Amounts");
```
Tickers -> Ticker
Full Names -> Full Name
Amounts -> Amount

quiz3 replied:
You have a point. I will change that.

awesominat commented on src/main/java/view/DashboardView.java:
```diff
@@ -81,93 +96,201 @@ public DashboardView(DashboardViewModel dashboardViewModel,
+        HashMap<String, String> userStats = state.getUserStats();
+        if (userStats != null) {
+//            userStatsTable.setPreferredSize(new Dimension(100, 200));
+            userStatsTable.setModel(new TableModel(userStats));
```
this displays daysSinceLastTopup = -1.0
first, change all the keys to English, so "Days Since Last Topup" and make it so if the days since last topped up is <0, then display "Never Topped Up".

balance -> Balance,
totalAssets -> Total Assets (maybe a short description if you want)

quiz3 replied:
Bruh, changing those keys is going to be so much work... but I guess we got to do it.

Changing those names which are buried behind maps is going to be a hassle but I suppose it is indeed worth it.

awesominat reviewed (changes requested):
great work so far. this is the main screen, so we have to make sure it looks the most polished

gursi26 commented:
Total assets should be total balance + total worth of all stocks owned, not just total worth of all stocks owned

quiz3 commented:
> great work so far. this is the main screen, so we have to make sure it looks the most polished

You're totally right, I just needed to get this thing pushed because I had already spent 3+ hours working on it and the law of diminishing returns was beginning to come into play in full force.

quiz3 commented:
> Total assets should be total balance + total worth of all stocks owned, not just total worth of all stocks owned

Wait but I thought the whole point of having assets separate from balance is that they are different things. If we wanted like total net worth, maybe what you said would make sense. 

I will need more clarification on this.

quiz3 commented on src/main/java/view/DashboardView.java:
```diff
@@ -81,93 +96,201 @@ public DashboardView(DashboardViewModel dashboardViewModel,
+//        userStatsTable.setMinimumSize(new Dimension(100, 100));
+
+        // Create new default table model for displaying stock information.
+        DefaultTableModel tableModel = new DefaultTableModel();
```
I am aware that I created the tables in the initializer as opposed to inside of the `propertyChange` method, as it is done in the other Views. 

I figured that because the only time the tables should change is when the use case is called, they should not be in the property change but instead in the initializer. The other views do not have a refresh button like the dashboard does, so I figured this made a difference and thus warranted different implementation.

rickypramanick replied:
I think its better to have this in the `propertyChange` method, I think its better if 'refresh' automatically once a sell/buy/reset has been performed rather than have the user have to manually do so. I understand that this may require a little more work but I think it would greatly simplify the look of the view and the user experience we provide.

rickypramanick replied:
If we do this, I reckon we scrap the refresh button.

gursi26 replied:
If it is in the initializer, **it only runs once**, regardless of whatever button is pressed.

If it it something that changes after we hit the run code button on IntelliJ, it has to be in the propertyChanged function

quiz3 commented on src/main/java/view/DashboardView.java:
```diff
@@ -81,93 +96,201 @@ public DashboardView(DashboardViewModel dashboardViewModel,
+        DefaultTableModel tableModel = new DefaultTableModel();
+
+        // Add columns to the model
+        tableModel.addColumn("Tickers");
+        tableModel.addColumn("Full Names");
+        tableModel.addColumn("Amounts");
```
I know what you are talking about from looking at the other use cases and views, but what you are suggesting is a complex re-working of the use case very late in the game. My assigned task was only ever officially to work on the view, so while I would be open to implementing this change, given my numerous other duties at the moment and the original delineation of tasks amongst group members, I would submit that this change could be implemented by any group member who wishes to help put the cherry on top of our main view and its use case.

quiz3 replied:
Use `getLastSeenPrice` from `Stock` entity to achieve this.

gursi26 replied:
I can work on it once the other issues on this pr are solved. Since i have a fairly clear idea of what needs to be done, I think I can do it fairly quickly.

gursi26 commented:
There should be three attributes then, 
- Current balance
- Portfolio net worth
- Total Assets (portfolio net worth + current balance)

The total assets attribute is very useful since it allows the user to see if the money they have has grown from the initial 10,000 that they were given on startup/reset.

quiz3 commented:
> Total assets should be total balance + total worth of all stocks owned, not just total worth of all stocks owned

Instead of this, we simply decided to change the name of `"totalAssets"` to be more clear about what it represents.

quiz3 committed:
Multiple featurs of DashboardView improved as requested by reviewers

quiz3 committed:
DashboardView's features have now been cleaned up as requested by reviewers

quiz3 committed:
merge conflicts resolved

quiz3 committed:
Merge remote-tracking branch 'origin' into dashboard-view-tables
merging with main pls work.

quiz3 commented:
Ok guys I know it passes Maven but this PR actually doesn't work so maybe wait on Approving it for the time being. Feel free to request more reviews. I want to know if I addressed all the concerns raised after the initial PR

quiz3 committed:
Invisible Tables bug fixed

gursi26 commented:
Neither table (user stats or stocks owned) updates after making a purchase.

If you want, I could take over and fix this (as well as add the features mentioned in #137 )

quiz3 commented:
> Neither table (user stats or stocks owned) updates after making a purchase (just tested it out locally).
> 
> 
> 
> If you want, I could take over and fix this (as well as add the features mentioned in #137 )

I thought they were only supposed to refresh upon the pressing of the refresh button. Are you suggesting they should refresh every time someone presses the back button from the Buy or Sell views?

#137 confuses me because I thought all that had been added. I added the market price of each owned stock to the state and display tables already. I suppose the price differential is not there yet. I would suggest that we don't need to display the original price but if it were in the state, we could certainly display the +/- differential between when they bought and current time. 

And yes, I would appreciate very much if you could take over for these most recently described changes, as they have to do with entities and use cases I am largely unfamiliar with. 

Thank you!

gursi26 committed:
fixed dashboard view and refresh button, displayed more stocks stats, temporarily removed a few user stats, still investigating bug with buy

gursi26 committed:
added days since last topup, portfolio net worth and net worth to user stats

gursi26 commented:
Summary of changes and what the current state of dashboard is;
- Added a new entity `StockInformation` that allows for the fetching of more stock information than just the price (which was the case with `PricePoint`). `StockInformation` now contains the current price, price change and percent change in price, all of which can be fetched in a single API call.
- Modified the `Finnhub` driver API class to have a method that returns a `StockInformation` object.
- Added the aforementioned statistics (current price, price change, percent difference) to the owned stocks table in the Dashboard view.
- The owned stocks table on the dashboard contains two types of information; information that is stored locally, such as the stocks owned by the user and the amounts of those stocks (and the entire userStats table), and information that needs to be fetched with an API call, such as the current price of those stocks, price difference and percent difference in stock price.
- Local information is refreshed whenever a back button is pressed from any other page. This way, when buying a new stock and returning to the dashboard, the new stock can be immediately seen in the owned stocks table on the dashboard. However, it's current price, price change and percent difference are set to 0 until the refresh button is hit.
- The refresh button causes an API call to fetch updated stock information from the API and update the table with new price information.

awesominat reviewed (changes requested):
it still shows -1 days since last reset balance.
also, the price change and percent change are nice, but not put into perspective, so they mean nothing on their own unless connected with a time period. (as in, price changes over 24h)

this is because, at first, the price change and percent change appear related to the stock held by the user (as in, their stock has changed 5% since last bought or something similar), but after investigation, it is clear that these are general stock information taken from the api.

please clarify this in the UI and add the price the user bought the stock at, as this is a crucial part of the whole app: to see how much the user has profited/lost thus far.

quiz3 commented:
> it still shows -1 days since last reset balance. also, the price change and percent change are nice, but not put into perspective, so they mean nothing on their own unless connected with a time period. (as in, price changes over 24h)
> 
> this is because, at first, the price change and percent change appear related to the stock held by the user (as in, their stock has changed 5% since last bought or something similar), but after investigation, it is clear that these are general stock information taken from the api.
> 
> please clarify this in the UI and add the price the user bought the stock at, as this is a crucial part of the whole app: to see how much the user has profited/lost thus far.

Bro, I fixed this though. The days since last topUp would fill in a string saying never topped up if that was indeed the case.

gursi26 commented:
I temporarily removed the previous code, I can refix the days since last topup issue.

Also, yes the stock price change and percent change are general stock information.

It is possible that the user purchased different amounts of the stock at different times, so putting a "price purchased at" column may be misleading.

awesominat commented:
round balance and net worth to .2f

these are price and percent changes over 24h. given that our app has no indication of whether the user is profiting, it is assumed that these changes are the only indication of progress, but they are not instead, they are simply general information. I am not saying other websites don't have this information, nor am I arguing that they mean nothing. I am arguing that they mean nothing *on their own*, as in, we need an indication of progress.

my issue is that the user has no way to know if they are profiting, they only have basic information like current price, d, and dp. our app means nothing if we don't tell the user if they are profiting or not, its just an addition / subtraction calculator at that point.

multiple online stock websites allow you to see the price at which you have bought the stocks.

you can achieve this by either keeping a running count of the $ the user has spent on a certain stock alongside the amount of it they own, or by looping through the transaction history and calculating it that way (I believe the second method is flawed, so the first one has to be the one)

gursi26 commented:
What is the meaning of running count? Do you mean running average or sum? 

Average makes more sense to me. We can loop over every buy transaction in the transaction history to compute the average buy price of a given stock. The average would need to reset when the user sells all units of a given stock that they own.

Also we do have information that allows the user to see how they are performing. net worth column of the user stats table is an example, where any number higher than 10,000 would indicate a profit. Perhaps per stock metrics would be helpful, but I do not think that was the point of this pr anyway, and it would be better off opened as a separate pr.

awesominat commented:
by running count, I mean running sum. they bought 2 stock A for $2 and then bought 2 the next day for $2.3, the running sum would be $4.3. contrast it with the next day where the price is 1.3 per stock, we can show the user that they have profited $0.3 or that they bought it for all of it for $4.3 and its current price is $5.2). I am ok with postponing it, but I feel like this is a missing crucial feature

awesominat reviewed (approved):
great work.

an issue detailing the lack of progress the user has should be created

gursi26 commented:
Ok, I will fix the issues with days since last top up and decimal places in balance and net worth before merging.

gursi26 committed:
fixed days since last reset showing -1.0, made everything 2 decimal places, new stocks now say 'Refresh to Update' in their price stats immeadiately after bought

gursi26 commented:
issues fixed, merging
